javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,245 @@
+    val channel2 = deferredChannel2.await()
+    assertThat(channel2).isNotSameInstanceAs(channel1)
+  }
+
+  @Test
+  fun invalidStoreOptionsFail() = runBlockingTest {
+    val storageServiceNg = createStorageService(this)
+    val e = assertFailsWith<AssertionError> {
+      storageServiceNg.openStorageChannel(
+        "fake store options".toByteArray(),
+        channelCallback,
+        messageCallback
+      )
+    }
+    assertThat(e).hasMessageThat()
+      .isEqualTo("ShadowParcel unable to unmarshall its custom format")
+  }
+
+  @Test
+  fun openStorageChannel_createsAStore() = runBlockingTest {
+    val storageServiceNg = createStorageService(this) as StorageServiceNgImpl
+
+    val storeOptions = StoreOptions(RamDiskStorageKey("store1"), CountType())
+    val encodedStoreOptions = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    assertThat(storageServiceNg.stores.size).isEqualTo(0)
+    storageServiceNg.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
+  }
+
+  @Test
+  fun openingTwoChannels_forDifferentStores_createsTwoStores() = runBlockingTest {
+    val deferredChannel1 = CompletableDeferred<IStorageChannel>()
+    val deferredChannel2 = CompletableDeferred<IStorageChannel>()
+
+    val storageServiceNg = createStorageService(this) as StorageServiceNgImpl
+    assertThat(storageServiceNg.stores.size).isEqualTo(0)
+
+    // Define the store options for two different stores.
+    val storeOptions1 = StoreOptions(RamDiskStorageKey("store1"), CountType())
+    val encodedStoreOptions1 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions1, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    val storeOptions2 = StoreOptions(RamDiskStorageKey("store2"), CountType())
+    val encodedStoreOptions2 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions2, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    // Open a storage channel for one store and assert that a new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions1,
+      createChannelCallback(deferredChannel1),
+      messageCallback
+    )
+    val channel1 = deferredChannel1.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
+
+    // Open a storage channel for another store and assert that another new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions2,
+      createChannelCallback(deferredChannel2),
+      messageCallback
+    )
+    val channel2 = deferredChannel2.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(2)
+
+    // Also check that the two channels that were created are distinct.
+    assertThat(channel2).isNotSameInstanceAs(channel1)
+  }
+
+  @Test
+  fun openingTwoChannels_forSameStore_onlyCreatesOneStore() = runBlockingTest {
```

Comment:
Do we want this assertion at all? (See comment above on not testing implementation details).

This is related to some behavior that we want: that we want two channels to interact with the same store. But I wonder if there's a way to verify that without checking the stores map?